### PR TITLE
do not assume event field values as mutable

### PR DESCRIPTION
**Pull request by colinsurprenant**
this PR build on top of #59 

- fix event mutability asumption.

relates to elastic/logstash#4264 and elastic/logstash#4191
requires elastic/logstash#4279 to pass specs under Java Event

### CODE REVIEW

**Comment by colinsurprenant** [comment]
@guyboertje friendly review ping! :D

File: lib/logstash/filters/mutate.rb [hunk 1]
```diff
@@ -428,12 +428,13 @@ def merge(event)
           next
         end
 
+        # No need to test the other
         if dest_field_value.is_a?(Hash)
-          # No need to test the other
-          event[dest_field].update(added_field_value)
+          # do not use event[dest_field].update to avoid assuming the in-place mutability of field values
```
Comment (jordansissel) [inline lib/logstash/filters/mutate.rb]: Maybe rephrase it somehow indicating that field values cannot be mutated in-place, only assigned. I'm not sure how to phrase it, but the current phrasing feels flimsy. We're moving to a model where an event field value 'get' will return a copied (or immutable) value, right?

**Comment by jordansissel** [comment]
Code LGTM. Left one note about phrasing/confusion on the comments added in the code. I'm not sure how to resolve, but will leave it to your judgement :)

File: lib/logstash/filters/mutate.rb (hunk 1 above)
Comment (colinsurprenant) [inline lib/logstash/filters/mutate.rb]: right, but not *immutable* strictly speaking, you can mutate it but it would be updated in Event itself...

**Comment by colinsurprenant** [comment]
@jordansissel let me know if comment wording is more comprehensible

**Comment by jordansissel** [comment]
More clear now, thank you!

LGTM

**Comment by colinsurprenant** [comment]
merged in master

**Comment by colinsurprenant** [comment]
pushed version 2.0.3, also include changes from PR #59
